dali/image/jpeg.cc

```diff
@@ -94,9 +96,13 @@ bool get_jpeg_size(const uint8 *data, size_t data_size, int *height, int *width)
 }  // namespace
 
 bool CheckIsJPEG(const uint8 *jpeg, int) {
+// Return false to indicate HostDecoder to use OpenCV fallback
+// see also dali/pipeline/operators/decoder/host_decoder.h:50
+#ifdef DALI_USE_JPEG_TURBO
```

Comment:
Gating this doesn't make sense -- the code doesn't depend on TJPG at all.

Reply:
As explained in comment: 
In dali/pipeline/operators/decoder/host_decoder.h:50, "CheckIsJPEG" is called to decide whether to use TJPG, by returning false decoder will fallback to OpenCV decoding

Reply:
That's not the case exactly -- if `CheckIsJPEG` returns true, it'll _try_ to use TJPG for decode, then if it fails it'll fall back and use OCV. You've already made changes to jpeg.cc, it'd make more sense to leave `CheckIsJPEG` as-is (after all, we're not fundamentally changing what is or isn't a JPEG file), and make it call opencv immediately instead of as a fallback.

Reply:
What I see in host_decoder.h is that return value of "DecodeJPEGHost" is handled by marco "DALI_CALL", and OCV fallback is in another "if" branch, chosen by "CheckIsJPEG" and "CheckIsPNG". 

Could you please explain it in more detail?

Reply:
Ok, so [host_decoder.h](https://github.com/NVIDIA/DALI/blob/master/dali/pipeline/operators/decoder/host_decoder.h#L50-L55) checks on the type of image being decoded. This then calls [DecodeJPEGHost](https://github.com/NVIDIA/DALI/blob/master/dali/image/jpeg.cc#L108) to perform the actual decode.

`DecodeJpegHost` then attempts to decode using TJPG, and if that fails, decodes using OCV. So if TJPG were to be disabled in the build, I would expect `DecodeJPEGHost` to still be called, and then that method to just call OCV. Make sense?

Reply:
Well, I mean currently, in [host_decoder.h](https://github.com/NVIDIA/DALI/blob/master/dali/pipeline/operators/decoder/host_decoder.h#L50-L55), an exception is thrown if ```DecodeJpegHost``` returns a failure, rather than retry decoding via OCV.

So should I change the strategy that using OCV as fallback if Jpeg or Png decoding fails?

Reply:
That's because DecodeJPEGHost (and DecodePNGHost, and any others we may add later) is always expected to succeed on a valid input. It maintains its own fallback path (in this case also OCV) in case TJPG cannot decode the image (or in your case, isn't available).

The code path you're relying on now is for image formats we don't have explicit paths for -- even if TJPG isn't available, we would still want the DecodeJPEGHost path used, and `CheckIsJPEG` to return true if the input is in fact a JPEG.

Reply:
So if I get you right, I should add OCV decoding codes to ```DecodeJpegHost``` in [host_decoder.h](https://github.com/NVIDIA/DALI/blob/master/dali/pipeline/operators/decoder/host_decoder.h) which is enabled when TJPG isn't available. 

I can do that.

Reply:
Yes please :)